packages/dashboard/src/components/queryEditor/iotSiteWiseQueryEditor/assetModelDataStreamExplorer/assetModelDataStreamExplorer.tsx

```diff
@@ -0,0 +1,246 @@
+
+  // const widgetTypesInSelection = selection?.types ?? [];
+  // xy-plot widgets are styled query objects
+  const t = selection?.type;
+
+  const [propertiesMaybe, setProperties] =
+    selection?.useProperty(
+      (properties) => properties,
+      (_properties, updatedProperties) => updatedProperties
+    ) ?? [];
+  const defaultQuery: QueryProperties = {
+    queryConfig: {
+      source: 'iotsitewise',
+      query: undefined,
+    },
+  };
+  const properties = propertiesMaybe ? maybeWithDefault(defaultQuery, propertiesMaybe) ?? defaultQuery : defaultQuery;
+
+  const setAssetModels = (updatedAssetModels: AssetModelQuery[] | undefined) => {
+    if (!setProperties || !t || !isJust(t)) return;
+
+    let updatedProperties = { ...properties };
+    const compositeWidgetForAggregationInformation = {
+      type: t.value,
+      properties,
+    } as QueryConfigWidget;
+    // hangle styled widget
+    if (t.value === 'xy-plot') {
+      const styledQuery = styledQueryWidgetOnDrop(
+        { assetModels: updatedAssetModels },
+        compositeWidgetForAggregationInformation
+      );
+      updatedProperties = {
+        ...properties,
+        queryConfig: {
+          ...properties.queryConfig,
+          query: {
+            ...properties.queryConfig.query,
+            assetModels: (styledQuery as unknown as IoTSiteWiseDataStreamQuery).assetModels,
+          },
+        },
+      };
+    } else {
+      const styledProperties = assignDefaultStyles({
+        ...compositeWidgetForAggregationInformation,
+        properties: {
+          ...compositeWidgetForAggregationInformation.properties,
+          queryConfig: {
+            ...compositeWidgetForAggregationInformation.properties.queryConfig,
+            query: {
+              ...compositeWidgetForAggregationInformation.properties.queryConfig.query,
+              assetModels: updatedAssetModels,
+            },
+          },
+        },
+      }).properties;
+      updatedProperties = {
+        ...styledProperties,
+      };
+    }
+
+    setProperties(updatedProperties);
+  };
+
+  const definedAssetModels = properties.queryConfig.query?.assetModels ?? [];
+
+  const onReset = useCallback(() => {
+    const clearedModelBasedWidgets = widgets.map(({ properties, ...rest }) => ({
+      ...rest,
+      properties: {
+        ...properties,
+        queryConfig: {
+          ...properties.queryConfig,
+          query: {
+            ...properties.queryConfig.query,
+            assetModels: [],
+          },
+        },
+      },
+    }));
+
+    update(clearedModelBasedWidgets);
+  }, [update, widgets]);
+
+  const onUpdateSelectedAsset = useCallback(
+    (updatedSelectedAsset?: AssetSummary) => {
+      if (updatedSelectedAsset && updatedSelectedAsset.id !== undefined) {
+        const id = updatedSelectedAsset.id;
+        const updatedSelectedAssets = widgets.map(({ properties, ...rest }) => ({
+          ...rest,
```

Comment:
may be worth writing a utility to update a deep nested object ? would be easier to understand ?

Reply:
Maybe we could have a utility to update assetModels for a widget, but this is the only place we are mutating assetModels right now so I don't think we need to make a fancy util for it.